### ISSUE TITLE
[Bug] - Mirroring External Types with multiple files - generates Redundant types in the dart code

### ISSUE THREAD

**Issue by dinbtechit** [fzyzcjy/flutter_rust_bridge#987]
### Describe the bug

Mirroring External Types with multiple files - generates Redundant types in the dart code

**my_types.rs**
```rs
#[frb(mirror(Configuration))]
#[derive(Default, Clone)]
pub struct _Configuration {
    pub id: u32,
    pub name: String
}
```

**api1.rs**
```rs
pub fn update_configuration(config: Configuration) -> anyhow::Result<()> {
   
}
```

**api2.rs**
```rs
pub fn perform_some_operaton(config: Configuration) -> anyhow::Result<()> {
   
}
```

```rs
// build.rs contains all rust input and output files
cargo build
```

### Codegen logs with `RUST_LOG=debug` environment variable

```shell
Finished dev [unoptimized + debuginfo] target(s) in 10.40s
```

### Generated binding code

Api1.dart

```dart
abstract Api1Class {
 . . .
}
 class Configuration {     <----- this class is generated in both Api1 and Api2
   . . .
 } 

class Api1ClassImpl implements Api1Class {
 
   . . . 
  Future<void> updateConfiguration (Configuration config) {

  }
  
}
```

**Api2.dart**
```dart
abstract Api2Class {
 . . .
}
 class Configuration {     <----- this class is generated in both Api1 and Api2
   . . .
 } 

class Api2ClassImpl implements Api2Class {
 
   . . . 
  Future<void> performSomeOtherAction(Configuration config) {
    . . .
  }
   
  
}
```

### OS

MacOS

### Version of `flutter_rust_bridge_codegen`

1.61.*

**Comment by fzyzcjy** [comment]
Looks like a bug. cc @dbsxdbsx who implemented the multi-file feature :)

**Comment by dbsxdbsx** [comment]
> Looks like a bug. cc @dbsxdbsx who implemented the multi-file feature :)

Just as I stated [here](https://github.com/fzyzcjy/flutter_rust_bridge/issues/956#issuecomment-1381778631), I wonder how to handle the customized struct.   
Since the customized struct would be used in both rust and dart code over different API blocks(files), duplicately defining them by adding suffixes to the customized struct name is not a good idea. Then, if this kind of structs is defined only once, where to place them?   Putting them in an auto-generated file say, `auto_genearated_shares.rs`?  If following this way, is `auto_genearated_shares.dart` needed?

**Comment by dinbtechit** [comment]
This is probably an issue with the custom struct as well I have not tested it. But I would expect FRB would maintain the same file structure as it was defined in the rust.

For example: I expect FRB to generate 3 dart classes for the case explained above: 
```
- `my_type.dart` (I expect this to have the Configuration class. Because it is where it is defined in the rust code.)
- `api1.dart` --|
                |-- These 2 files should just import the Configuration class from my_type.dart
- `api2.dart` --|
```

**Comment by dbsxdbsx** [comment]
> > Looks like a bug. cc @dbsxdbsx who implemented the multi-file feature :)
> 
> Just as I stated [here](https://github.com/fzyzcjy/flutter_rust_bridge/issues/956#issuecomment-1381778631), I wonder how to handle the customized struct.
> 
> Since the customized struct would be used in both rust and dart code explicitly by users over different API blocks(files), I think duplicately defining them by adding different suffixes to the original name of the customized struct is not a good idea.
> 
> Then, in the context of FFI, if this kind of structs is defined only once, where to place them? Putting them in a new auto-generated file say, `auto_generated_shares.rs`? If following this way, is `auto_generated_shares.dart` needed, too, where contains the definition of the structs for usage in dart?

@fzyzcjy, I am currently trying to reconstruct the code to make everything shared (not only customized struct) into another auto-generated file, say `bridge_generated_shared(_io/_wasm).rs/dart`. If this is done, there would be no redundant generated code for each regular API block, and there is no need to add suffixes for some shared auto-generated structs, like `wire_uint_8_list`  would only be defined in an auto-generated shared file, something like `wire_uint_8_list_0/1....` would never occur then. 

Though the idea is simple, this does need to refractor lots of code.  If this is not a proper roadmap for frb, please inform me ASAP.

**Comment by fzyzcjy** [comment]
@dbsxdbsx That looks reasonable to me generally.

One concern may be that, will it make code complicated a lot? If so, I am not sure whether it is worth it, since the maintenance cost of frb will be much higher.

Also cc @Desdaemon @rogurotus who made a lot of contributions to frb - what do you think?

**Comment by rogurotus** [comment]
If I understand correctly, then we are talking about generating common classes for the same types from different api_*.rs.
It's definitely worth it, since due to the current behavior, you have to write +- all api in one api.rs because of such collisions. But it does feel like it won't be easy to do so.

**Comment by dbsxdbsx** [comment]
> If I understand correctly, then we are talking about generating common classes for the same types from different api_*.rs. It's definitely worth it, since due to the current behavior, you have to write +- all api in one api.rs because of such collisions. But it does feel like it won't be easy to do so.

Roughly correct. Currently, if a type is used in more than one API block(rust/dart file), the type would be defined respectively with different suffixes in each used API block, otherwise, there would be an issue, like this post. Now I am gonna extract them into a shared auto-generated block, where there is only shared stuff.

**Comment by fzyzcjy** [comment]
Btw, it would be great if this logic can be (mostly) put into a separate mod, i.e. encapsulation or "Low Coupling, High Cohesion". For example, I guess this separate mod may have a brief API like "fn query_which_file_should_this_thing_be_put_into()" etc. Then the code may be clean and not hard to maintain.

**Comment by dbsxdbsx** [comment]
> Btw, it would be great if this logic can be (mostly) put into a separate mod, i.e. encapsulation or "Low Coupling, High Cohesion". For example, I guess this separate mod may have a brief API like "fn query_which_file_should_this_thing_be_put_into()" etc. Then the code may be clean and not hard to maintain.
Maybe it is time to write some code to first test if this idea is technically feasible, then try to make the code as  "Low Coupling, High Cohesion" as possible.

**Comment by fzyzcjy** [comment]
Sure!

**Comment by dbsxdbsx** [comment]
> > If I understand correctly, then we are talking about generating common classes for the same types from different api_*.rs. It's definitely worth it, since due to the current behavior, you have to write +- all api in one api.rs because of such collisions. But it does feel like it won't be easy to do so.
> 
> Roughly correct. Currently, if a type is used in more than one API block(rust/dart file), the type would be defined respectively with different suffixes in each used API block. Otherwise, there would be an issue, like this post. Now I am planning to extract them all into a shared auto-generated block, where there is only shared stuff. And with this mechanism, when the shared type is used from the regular API block, it is called from the shared (auto-generated) block, be the type auto-generated or customized by users.

Now I notice that this shared refractor is essential to solving https://github.com/fzyzcjy/flutter_rust_bridge/issues/956. Otherwise, for a customized structure defined not in any API block, frb would be at a loss on where to put the methods(APIs) defined in `impl`.

**Comment by svenstaro** [comment]
This is quite relevant and prohibits complex setups right now.

**Comment by franklyonnet** [comment]
FYI, this is how we deal with it right now. We are using the CLI version of the codegen and we are post-processing with this:

use regex::Regex;
use std::fs::{self, OpenOptions};
use std::io::{Read, Write};
use std::path::PathBuf;
use std::collections::{HashMap, HashSet};

const C_MACOS_OUTPUT_PREFIX: &str = "./macos/Runner/bridge_generated_";
const C_IOS_OUTPUT_PREFIX: &str = "./ios/Runner/bridge_generated_";
const FILE_COUNT: usize = 6;
const MODULES: [&str; FILE_COUNT] = [
    "xxx",
    "xxx2",
    "xxx3",
    "xxx4",
    "xxx5",
    "xxx6",
];


fn create_string_vector(prefix: &str, modules: &[&str], suffix: &str) -> Vec<String> {
    modules.iter().map(|module| format!("{}{}.{}", prefix, module, suffix)).collect()
}

fn remove_duplicate_structs(file_paths: &[String]) -> std::io::Result<()> {
    let struct_pattern = Regex::new(r"(?s)typedef struct\s+[\w_]+\s*\{.*?\}\s*[\w_]+;").unwrap();
    let mut seen_structs = HashMap::new();

    for file_path in file_paths {
        let path = PathBuf::from(file_path);
        let mut file_content = String::new();
        {
            println!("Reading file: {}", file_path);
            let mut file = fs::File::open(&path)?;
            file.read_to_string(&mut file_content)?;
        }

        let mut new_file_content = String::new();
        let mut last_end = 0;

        for struct_def in struct_pattern.find_iter(&file_content) {
            let start = struct_def.start();
            let end = struct_def.end();
            let struct_name = struct_def.as_str().split_whitespace().nth(2).unwrap();

            if !seen_structs.contains_key(struct_name) {
                seen_structs.insert(struct_name.to_string(), ());
                new_file_content.push_str(&file_content[last_end..end]);
            } else {
                new_file_content.push_str(&file_content[last_end..start]);
            }

            last_end = end;
        }

        new_file_content.push_str(&file_content[last_end..]);

        let mut file = OpenOptions::new()
            .write(true)
            .truncate(true)
            .open(&path)?;
        file.write_all(new_file_content.as_bytes())?;
    }

    Ok(())
}

fn main() {

    let c_macos_output = create_string_vector(C_MACOS_OUTPUT_PREFIX, &MODULES, "h");
    let c_ios_output = create_string_vector(C_IOS_OUTPUT_PREFIX, &MODULES, "h");

    if let Err(e) = remove_duplicate_structs(&c_macos_output) {
        eprintln!("Error when removing duplicates: {}", e);
    }

    if let Err(e) = remove_duplicate_structs(&c_ios_output) {
        eprintln!("Error when removing duplicates: {}", e);
    }
}

**Comment by dinbtechit** [comment]
Please remove wontfix label.

**Comment by svenstaro** [comment]
Has this actually been fixed?

**Comment by fzyzcjy** [comment]
No, it is the bot which automatically close it :/

**Comment by dbsxdbsx** [comment]
> Has this actually been fixed?

Sorry for that, I am still working on it.

**Comment by fzyzcjy** [comment]
v2 supports multiple files (indeed support a whole crate with arbitrary files and folders), so I will close this, but feel free to reopen!
